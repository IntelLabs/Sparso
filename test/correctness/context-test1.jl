#=
Copyright (c) 2015, Intel Corporation

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of Intel Corporation nor the names of its contributors
      may be used to endorse or promote products derived from this software
      without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
=#

include("../../src/SparseAccelerator.jl")
using SparseAccelerator

set_options(SA_ENABLE, SA_VERBOSE, SA_USE_SPMP, SA_CONTEXT, SA_REORDER, SA_REPLACE_CALLS)

#include("./pcg-symgs.jl")
include("utils.jl")

# TEMPORARY: this should be removed in future once structure analysis can
# recognize that L and U are part of A
# Copied from pcg-symgs.jl to avoid affecting other tests if it were changed.
function pcg_symgs(x, A, b, tol, maxiter)
    # These are the only 3 properties we need to discover automatically.
    # We also require SA_CONST_STRUCTURED for A, but that has already been discovered.
    set_matrix_property(:L, SA_LOWER_OF, :A)
    set_matrix_property(:U, SA_UPPER_OF, :A)
    set_matrix_property(Dict(
        :A => SA_SYMM_STRUCTURED | SA_SYMM_VALUED,
        :U => SA_CONST_VALUED
        )
    )

    total_time = time()

    trsv_time = 0.
    spmv_time = 0.
    blas1_time = 0.

    L = tril(A)
    U  = SparseMatrixCSC{Cdouble, Cint}(spdiagm(1./diag(A))*triu(A))

    spmv_time -= time()
    r = b - A * x
    spmv_time += time()

    blas1_time -= time()
    normr0 = norm(r)
    z = copy(r)
    blas1_time += time()

    rel_err = 1

    trsv_time -= time()
    Base.SparseMatrix.fwdTriSolve!(L, z)
    Base.SparseMatrix.bwdTriSolve!(U, z)
    trsv_time += time()

    blas1_time -= time()
    p = copy(z) #NOTE: do not write "p=z"! That would make p and z aliased (the same variable)
    rz = dot(r, z)
    blas1_time += time()

    # For statistics only. We get these values so that L, U and A won't be live out of
    # the loop -- to save reverse-reordering of them.
    nnzL = nnz(L)
    nnzU = nnz(U)
    sizeL1 = size(L, 1) 
    sizeL2 = size(L, 2)
    nnzA = nnz(A)
    sizeA1 = size(A, 1) 
    sizeA2 = size(A, 2)

    k = 1

    while k <= maxiter
        old_rz = rz

        spmv_time -= time()
        Ap = A*p # Ap = SparseAccelerator.SpMV(A, p) # This takes most time. Compiler can reorder A to make faster
        spmv_time += time()

        blas1_time -= time()
        alpha = old_rz / dot(p, Ap)
        x += alpha * p
        r -= alpha * Ap
        rel_err = norm(r)/normr0
        blas1_time += time()

        if rel_err < tol 
            break
        end

        blas1_time -= time()
        z = copy(r)
        blas1_time += time()

        trsv_time -= time()
        Base.SparseMatrix.fwdTriSolve!(L, z)
          # Could have written as z=L\z if \ is specialized for triangular
        Base.SparseMatrix.bwdTriSolve!(U, z)
          # Could have wrriten as z=U\z if \ is specialized for triangular
        trsv_time += time()

        blas1_time -= time()
        rz = dot(r, z)
        beta = rz/old_rz
        p = z + beta * p
        blas1_time += time()

        k += 1
    end

    total_time = time() - total_time
    println("total = $(total_time)s trsv_time = $(trsv_time)s ($((12.*(nnzL   + nnzU  ) + 2.*8*(sizeL1     + sizeL2    ))*k/trsv_time/1e9) gbps) spmv_time = $(spmv_time)s ($((12.*nnzA   + 8.*(sizeA1     + sizeA2    ))*(k + 1)/spmv_time/1e9) gbps) blas1_time = $blas1_time")
#   println("total = $(total_time)s trsv_time = $(trsv_time)s ($((12.*(nnz(L) + nnz(U)) + 2.*8*(size(L, 1) + size(L, 2)))*k/trsv_time/1e9) gbps) spmv_time = $(spmv_time)s ($((12.*nnz(A) + 8.*(size(A, 1) + size(A, 2)))*(k + 1)/spmv_time/1e9) gbps) blas1_time = $blas1_time")
 
    return x, k, rel_err
end

if length(ARGS) == 0
    m = 10
    A = generate_symmetric_nonzerodiagonal_sparse_matrix(m)
else
    A = matrix_market_read(ARGS[1], true, true)
    m = size(A, 1)
end
x       = zeros(Float64, m)
b       = ones(Float64, m)
tol     = 1e-7
maxiter = 20000

x, k, rel_err = pcg_symgs(x, A, b, tol, maxiter)
x = zeros(Float64, m)
println("Original: ")
x, k, rel_err = pcg_symgs(x, A, b, tol, maxiter)
println("\tOriginal k=", k)
println("\tOriginal rel_err=", rel_err)

A2 = copy(A) # workaround that we change A in-place
@acc x, k, rel_err = pcg_symgs(x, A2, b, tol, maxiter)
println("\nAccelerated: ")
x = zeros(Float64, m)
@acc x, k, rel_err = pcg_symgs(x, A, b, tol, maxiter)
println("\tAccelerated k=", k)
println("\tAccelerated rel_err=", rel_err)
