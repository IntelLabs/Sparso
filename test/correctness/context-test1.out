Matrix structures discovered:
    Union{GenSym,Symbol}[:L,:A]

Constant structures discovered:
    Union{GenSym,Symbol}[:L,:A]

Value symmetry discovered:

Structure symmetry discovered:

Constants discovered:
    Set(Union{GenSym,Symbol}[:tol,:A,:normr0,:L,:maxiter])

Single-defs discovered:
    Set(Union{GenSym,Symbol}[:beta,:rel_err,:k,:rz,:Ap,:z,:p,:U,:alpha,:old_rz,:r,:x])


According to pattern
    CS_SpMV_pattern1
    (:call,:(Main.*),Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer},Array{T,1})
    (:NO_SUB_PATTERNS,)
    do_nothing
    (:call,:((top(getfield))(SparseAccelerator,:SpMV)),:arg2,:arg3)
    gather_context_sensitive_info
    NewFunctionKnob
    DeleteFunctionKnob
    (:arg2,)
    40
    (:arg2,:arg3)
replace
    Expr call [Array{Float64,1}]
        Main.* [GlobalRef]
        A::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32} [SymbolNode] [Base.SparseMatrix.SparseMatrixCSC{Float64,Int32}]
        p::Array{Float64,1} [SymbolNode] [Array{Float64,1}]

to
    Expr call [Array{Float64,1}]
        Expr call [Function]
            top(getfield) [TopNode]
            SparseAccelerator [Symbol] [Void]
            :SpMV [QuoteNode]
        A::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32} [SymbolNode] [Base.SparseMatrix.SparseMatrixCSC{Float64,Int32}]
        p::Array{Float64,1} [SymbolNode] [Array{Float64,1}]


Context-sensitive actions to take:
    Insert action:
        ##mknobA#8208 = (SparseAccelerator.new_matrix_knob)(A,true,true,false,false,false,false)
        ##fknob#8203 = (SparseAccelerator.new_function_knob)()
    outisde loop of BBs [0, 2, 1] before loop head BB 0
    
    Insert action:
        (SparseAccelerator.delete_matrix_knob)(##mknobA#8208)
        (SparseAccelerator.delete_function_knob)(##fknob#8203)
    on the edge BB 2 --> BB -5
    
    Insert action:
        (SparseAccelerator.delete_matrix_knob)(##mknobA#8208)
        (SparseAccelerator.delete_function_knob)(##fknob#8203)
    on the edge BB 0 --> BB -4
    
    Insert action:
        (SparseAccelerator.add_mknob_to_fknob)(##mknobA#8208,##fknob#8203)
    outisde loop of BBs [0, 2, 1] before loop head BB 0
    


New AST:
    $(Expr(:lambda, Any[:x,:A,:b,:tol,:maxiter], Any[Any[Any[:x,Array{Float64,1},18],Any[:A,Base.SparseMatrix.SparseMatrixCSC{Float64,Int32},0],Any[:b,Array{Float64,1},0],Any[:tol,Float64,0],Any[:maxiter,Int64,0],Any[:L,Base.SparseMatrix.SparseMatrixCSC{Float64,Int32},18],Any[:U,Any,18],Any[:M,Any,18],Any[:r,Array{Float64,1},2],Any[:normr0,Float64,18],Any[:rel_err,Any,2],Any[:z,Array{Float64,1},2],Any[:p,Array{Float64,1},2],Any[:rz,Float64,2],Any[:k,Int64,2],Any[:time1,Float64,18],Any[:old_rz,Float64,18],Any[:Ap,Array{Float64,1},18],Any[:alpha,Float64,18],Any[:beta,Float64,18]],Any[],Any[],Any[]], :(begin  # /media/sf_VBoxVMShared/SparseAccelerator/cleanup/test/correctness/./pcg-symgs.jl, line 2:
            NewvarNode(:old_rz)
            NewvarNode(:Ap)
            NewvarNode(:alpha)
            NewvarNode(:beta)
            L = (Main.tril)(A::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32})::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32} # line 3:
            U = (Main.spdiagm)(1 ./ (Main.diag)(A::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32})::Array{Float64,1}::Array{Float64,1})::Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer} * (Main.triu)(A::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32})::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32} # line 4:
            M = L::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32} * U # line 5:
            r = b::Array{Float64,1} - A::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32} * x::Array{Float64,1}::Array{Float64,1}::Array{Float64,1} # line 6:
            normr0 = (Main.norm)(r::Array{Float64,1})::Float64 # line 7:
            rel_err = 1 # line 9:
            z = (Main.copy)(r::Array{Float64,1})::Array{Float64,1} # line 10:
            ((top(getfield))(Base.SparseMatrix,:fwdTriSolve!)::F)(L::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32},z::Array{Float64,1})::Array{Float64,1} # line 11:
            ((top(getfield))(Base.SparseMatrix,:bwdTriSolve!)::F)(U,z::Array{Float64,1})::Array{Float64,1} # line 13:
            p = (Main.copy)(z::Array{Float64,1})::Array{Float64,1} # line 14:
            rz = (Main.dot)(r::Array{Float64,1},z::Array{Float64,1})::Float64 # line 15:
            k = 1 # line 16:
            time1 = (Main.time)()::Float64 # line 17:
            NewvarNode(:old_rz)
            NewvarNode(:Ap)
            NewvarNode(:alpha)
            NewvarNode(:beta)
            unless k::Int64 <= maxiter::Int64::Bool goto 3
            4: 
            ##mknobA#8208 = (SparseAccelerator.new_matrix_knob)(A,true,true,false,false,false,false)
            ##fknob#8203 = (SparseAccelerator.new_function_knob)()
            (SparseAccelerator.add_mknob_to_fknob)(##mknobA#8208,##fknob#8203)
            goto 0
            0:  # line 18:
            old_rz = rz::Float64 # line 19:
            Ap = ((top(getfield))(SparseAccelerator,:SpMV))(A::Base.SparseMatrix.SparseMatrixCSC{Float64,Int32},p::Array{Float64,1},##fknob#8203)::Array{Float64,1} # line 20:
            alpha = old_rz::Float64 / (Main.dot)(p::Array{Float64,1},Ap::Array{Float64,1})::Float64::Float64 # line 21:
            x = x::Array{Float64,1} + alpha::Float64 * p::Array{Float64,1}::Array{Float64,1}::Array{Float64,1} # line 22:
            r = r::Array{Float64,1} - alpha::Float64 * Ap::Array{Float64,1}::Array{Float64,1}::Array{Float64,1} # line 23:
            rel_err = (Main.norm)(r::Array{Float64,1})::Float64 / normr0::Float64::Float64 # line 24:
            unless rel_err::Float64 < tol::Float64::Bool goto 1
            (SparseAccelerator.delete_matrix_knob)(##mknobA#8208)
            (SparseAccelerator.delete_function_knob)(##fknob#8203) # line 25:
            goto 3
            1:  # line 28:
            z = (Main.copy)(r::Array{Float64,1})::Array{Float64,1} # line 29:
            nothing # line 31:
            nothing # line 34:
            rz = (Main.dot)(r::Array{Float64,1},z::Array{Float64,1})::Float64 # line 35:
            beta = rz::Float64 / old_rz::Float64::Float64 # line 36:
            p = z::Array{Float64,1} + beta::Float64 * p::Array{Float64,1}::Array{Float64,1}::Array{Float64,1} # line 37:
            k = k::Int64 + 1::Int64
            2: 
            unless (top(!))(k::Int64 <= maxiter::Int64::Bool)::Bool goto 0
            (SparseAccelerator.delete_matrix_knob)(##mknobA#8208)
            (SparseAccelerator.delete_function_knob)(##fknob#8203)
            3:  # line 39:
            return (top(tuple))(x::Array{Float64,1},k::Int64,rel_err::Union{Float64,Int64})::Tuple{Array{Float64,1},Int64,Union{Float64,Int64}}
        end::Tuple{Array{Float64,1},Int64,Union{Float64,Int64}})))
********************************************************************************